sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotNestTypesInArguments.cs

```diff
@@ -67,37 +71,52 @@ node switch
                 _ => Enumerable.Empty<ParameterSyntax>()
             };
 
-        private static int GetNestingDepth(ParameterSyntax parameterSyntax, SyntaxNodeAnalysisContext context) =>
-            GetNestingDepth(context.SemanticModel.GetDeclaredSymbol(parameterSyntax)?.Type, 0 , true);
+        private sealed class GenericWalker : CSharpSyntaxWalker
+        {
+            private static readonly ImmutableArray<KnownType> ignoredTypes =
+                KnownType.SystemFuncVariants
+                         .Union(KnownType.SystemActionVariants)
+                         .Union(new[] { KnownType.System_Linq_Expressions_Expression_T })
+                         .ToImmutableArray();
 
-        private static readonly ImmutableArray<KnownType> expressionFuncActionTypes =
-            KnownType.SystemFuncVariants
-            .Union(KnownType.SystemActionVariants)
-            .Union(new[] { KnownType.System_Linq_Expressions_Expression_T })
-            .ToImmutableArray();
+            private readonly int maxDepth;
+            private readonly SemanticModel model;
 
-        private static int GetNestingDepth(ITypeSymbol argumentSymbol, int depth, bool ignoreExpressions)
-        {
-            var currentOrNestedArgument = argumentSymbol;
-            if (currentOrNestedArgument is IArrayTypeSymbol arrayTypeSymbol)
-            {
-                currentOrNestedArgument = arrayTypeSymbol.ElementType;
-            }
+            private int depth;
+
+            public bool HasReachedMaxDepth { get; private set; }
 
-            if (!(currentOrNestedArgument is INamedTypeSymbol namedTypeSymbol))
+            public GenericWalker(int maxDepth, SemanticModel model)
             {
-                return depth;
+                this.maxDepth = maxDepth;
+                this.model = model;
             }
 
-            var nextDepth = ignoreExpressions && namedTypeSymbol.ConstructedFrom.IsAny(expressionFuncActionTypes) ? depth : depth + 1;
-
-            var typeArguments = namedTypeSymbol.TypeArguments;
-            if (typeArguments != null && typeArguments.Any())
+            public override void VisitGenericName(GenericNameSyntax node)
             {
-                return typeArguments.Max(p => GetNestingDepth(p, nextDepth, false));
-            }
+                if (!(model.GetSymbolInfo(node).Symbol is INamedTypeSymbol namedTypeSymbol))
+                {
+                    return;
+                }
 
-            return depth;
+                if (!namedTypeSymbol.ConstructedFrom.IsAny(ignoredTypes))
```

Comment:
I'd prefer positive condition here and swap if/else blocks. The `not is any` is hard :)